v2/api/dataprotection/customizations/backup_vaults_backup_instance_extensions.go

```diff
@@ -0,0 +1,142 @@
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	owner genruntime.MetaObject,
+	_ *resolver.Resolver,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger,
+	_ extensions.PostReconcileCheckFunc,
+) (extensions.PostReconcileCheckResult, error) {
+	log.V(Debug).Info("########################## Starting Post-reconcilation for Backup Instance ##########################")
+	backupInstance, ok := obj.(*dataprotection.BackupVaultsBackupInstance)
+	if !ok {
+		return extensions.PostReconcileCheckResult{},
+			errors.Errorf("cannot run on unknown resource type %T, expected *dataprotection.BackupVaultsBackupInstance", obj)
+	}
+
+	var _ conversion.Hub = backupInstance
+
+	if owner != nil && backupInstance != nil && backupInstance.Status.Id != nil &&
+		backupInstance.Status.Properties != nil && backupInstance.Status.Properties.ProtectionStatus != nil &&
+		backupInstance.Status.Properties.ProtectionStatus.Status != nil {
+
+		protectionStatus := *backupInstance.Status.Properties.ProtectionStatus.Status
+		log.V(Debug).Info(fmt.Sprintf("########################## Protection Status is  %q ##########################", protectionStatus))
+
+		// return success for reconcilation if the state is non-retryable
+		if nonRetryableStates.Contains(protectionStatus) {
+			log.V(Debug).Info("########################## Returning PostReconcileCheckResultSuccess ##########################")
+			return extensions.PostReconcileCheckResultSuccess(), nil
+		}
+
+		// call sync api only when protection status is ProtectionError and error code is usererror
+		if protectionStatus == "ProtectionError" {
+			protectionStatusErrorCode := ""
+			protectionStatusErrorCode = strings.ToLower(*backupInstance.Status.Properties.ProtectionStatus.ErrorDetails.Code)
+			log.V(Debug).Info(fmt.Sprintf("########################## Protection Error code is  %q ##########################", protectionStatusErrorCode))
+
+			if protectionStatusErrorCode != "" && strings.Contains(protectionStatusErrorCode, "usererror") {
+				id, _ := genruntime.GetAndParseResourceID(backupInstance)
+				subscription := id.SubscriptionID
+				rg := id.ResourceGroupName
+				vaultName := id.Parent.Name
+
+				log.V(Debug).Info("########################## Starting NewBackupInstancesClient for Backup Instance ##########################")
+				clientFactory, err := armdataprotection.NewClientFactory(subscription, armClient.Creds(), armClient.ClientOptions())
+				if err != nil {
+					log.Error(err, "failed to create armdataprotection client")
+				}
+
+				var parameters armdataprotection.SyncBackupInstanceRequest
+				parameters.SyncType = to.Ptr(armdataprotection.SyncTypeDefault)
+				log.V(Debug).Info("########################## Starting BeginSyncBackupInstance for Backup Instance ##########################")
+				poller, err := clientFactory.NewBackupInstancesClient().BeginSyncBackupInstance(ctx, rg, vaultName, backupInstance.AzureName(), parameters, nil)
```

Comment:
Beginning a sync on every reconciliation seems odd - shouldn't you be checking whether that's needed or not first? If the resource is already in the goal state, there's no reason to do this.

Reply:
thats alredy been done as part of above if block

if protectionStatus == "ProtectionError"